packages/block-editor/src/utils/object.js

```diff
@@ -93,24 +45,20 @@ function cloneObject( object ) {
  * @return {Object} Cloned object with the new value set.
  */
 export function setImmutably( object, path, value ) {
-	const normalizedPath = normalizePath( path );
-	const newObject = object ? cloneObject( object ) : {};
+	path = Array.isArray( path ) ? path : [ path ];
+
+	if ( path.length === 0 ) {
+		return value;
+	}
+
+	const shallowClone = Array.isArray( object )
+		? [ ...object ]
+		: { ...object };
+	const [ first, ...rest ] = path;
 
-	normalizedPath.reduce( ( acc, key, i ) => {
-		if ( acc[ key ] === undefined ) {
-			if ( Number.isInteger( path[ i + 1 ] ) ) {
-				acc[ key ] = [];
-			} else {
-				acc[ key ] = {};
-			}
-		}
-		if ( i === normalizedPath.length - 1 ) {
-			acc[ key ] = value;
-		}
-		return acc[ key ];
-	}, newObject );
+	shallowClone[ first ] = setImmutably( shallowClone[ first ], rest, value );
```

Comment:
Is there a way to achieve the same without recursivity? I think recursivity has a performance overhead that is best avoided in low level functions that get called often.

Reply:
Unlike `cloneObjects`, which was using a recursive function in for every object inside, this one is only called for every item in the path though. But I can check.

Reply:
I'm guessing it's wrong for nested arrays.

Reply:
The for loop is weird, it should be

```
	 for ( let i = 0; i < path.length - 1; i++ ) {
		const key = path[ i ];
		currentObj[ key ] = Array.isArray( currentObj[ key ] )
			? [ ...currentObj[ key ] ]
			: { ...currentObj[ key ] };
		currentObj = currentObj[ key ];
	}

Reply:
> I think recursivity has a performance overhead that is best avoided in low level functions that get called often.

did you have any indication of this? it seems surprising to me…unless we're talking about non-tail-recursive forms or things that might overflow the stack.

this one looks benign, but we can always try and see if there's an obvious impact. I'm not arguing we should stick to recursion; only asking because the notion was unexpected.

Reply:
I also find this a lot harder to read. The previous function had many many more nested function calls.

Reply:
The number of nested functions = `path.length`. Even if you'd do something like a forEach or reduce loop you'd have the same number of nested function. It really seems ok to me. We're not deep cloning here.

Reply:
> non-tail-recursive

yes, this is the main thing for me. I agree that it's harder to read and for this particular use-case, it might not be much recursivity (path is short in most cases), so I'm fine shipping this but I think in JavaScript specifically, it's best to avoid recursion. functions arguments get copied over and over again from caller to callee... and same for return values.

Reply:
Pushed 7bdcfd06047551c04afb820242a373af65295b98

Reply:
I tweaked it even further after that, I think it makes it more readable and avoids the if statement. Also cached `prev[ key ]`. Anyway at this point we're probably micro optimising for performance and readability. :)